manual/CONFIGURATION.md
# Configuration
<p align="center"><img width="300" src="https://cdn.rawgit.com/pixelass/paraglider/master/paraglider.svg" alt="logo"/></p>

<!-- toc -->

- [Class names](#class-names)
- [Options](#options)
  * [Loop](#loop)
  * [Visible slides](#visible-slides)
  * [Amount of next slides](#amount-of-next-slides)
  * [Enable Touch](#enable-touch)
  * [Enable Swipe](#enable-swipe)
  * [Initial slide](#initial-slide)
  * [Threshold](#threshold)
  * [Snap back](#snap-back)
    + [Speed and duration](#speed-and-duration)
  * [Add classes](#add-classes)
- [Callbacks](#callbacks)
  * [Initialized](#initialized)
  * [Destroyed](#destroyed)
  * [Sliding](#sliding)
  * [Finished](#finished)

<!-- tocstop -->

## Class names

* `{object} classNames`
  * Mapping of class names to be used by the plugin.
* `{string} [classNames.pluginLoaded='pluginLoaded']`
  * Applied when the plugin has been loaded
* `{string} [classNames.init='init']`
  * Applied when the plugin has been initialized. Removed on first interaction.
* `{string} [classNames.slides='slides']`
  * This element will be used to track touches. This is the wrapper around the slides.
* `{string} [classNames.slide='slide']`
  * Selector for each single slide.
* `{string} [classNames.current='current']`
  * Applied to the currently visible slide
* `{string} [classNames.previous='previous']`
  * Applied to the previous slide in the queue
* `{string} [classNames.next='next']`
  * Applied to the next slide in the queue
* `{object} addClasses`
  * Map of class names to be added to slides.
* `{object} [addClasses.previous=true]`
  * Sets previous class name to slides.
* `{object} [addClasses.current=true]`
  * Sets current class name to slides.
* `{object} [addClasses.next=true]`
  * Sets next class name to slides.
* `{object} addMultiClasses`
  * Map of class names to be added to slides using a counter.
* `{object} [addMultiClasses.previous=undefined]`
  * Sets previous class name to slides using a counter.
* `{object} [addMultiClasses.current]`
  * Sets current class name to slides using a counter.
* `{object} [addMultiClasses.next=undefined]`
  * Sets next class name to slides using a counter.


## Options

* `{boolean} [enableTouch=true]`
  * Enable touch interaction.
* `{boolean} [enableSwipe=true]`
  * Enable mouse/drag/swipe interaction.
* `{boolean} [loop=false]`
  * Enable looping slides.
* `{number} [speed=250]`
  * Animation duration when using paging.
* `{number} [spring=100]`
  * Animation duration when snapping.
* `{number} [snapBackAt=0.25]`
  * Amount of distance needed to snap. [0, 1]
* `{number} [threshold=10]`
  * Threshold of pixels until the sliding mechanisms is triggered.
* `{number} [initialSlide=0]`
  * Initially visible slide
* `{number} [visibleSlides=1]`
  * Amount of visible slides
* `{number} [slideBy=1]`
  * Amount of slides to slide on interaction

### Loop

`[loop=true] {boolean}`

Looping can be disabled with this flag.

> Warning! This does not work well when `visibleSlides > 1`.  
> Issue reports welcome  
> Contributions welcome  

```js
import {belt} from 'paraglider/presets'

belt(document.querySelector('.belt'), {loop: false})
```

### Visible slides

`[visibleSLides=1] {number}`

Show more than one slide. 

> Warning! This does not work well when `loop === false`.  
> Issue reports welcome  
> Contributions welcome  

```js
import {multiBelt} from 'paraglider/presets'

multiBelt(document.querySelector('.belt'), {visibleSlides: 3})
```

### Amount of next slides

`[slideBy=1] {number}`

Allows sliding by more than one item.

> Warning! Requires `visibleSlides >= slideBy`.  

```js
import {multiBelt} from 'paraglider/presets'

multiBelt(document.querySelector('.belt'), {visibleSlides: 3})
```

### Enable Touch

`[enableSwipe=true] {boolean}`

Disable or enable swiping with touch.

```js
import {belt} from 'paraglider/presets'

belt(document.querySelector('.belt'), {enableTouch: false})

```

### Enable Swipe

`[enableSwipe=true] {boolean}`

Disable or enable swiping with the mouse.

```js
import {belt} from 'paraglider/presets'

belt(document.querySelector('.belt'), {enableSwipe: false})

```

### Initial slide

`[initialSlide=0] {boolean}`

Set the initially active slide

```js
import {belt} from 'paraglider/presets'

belt(document.querySelector('.belt'), {initialSlide: 1})
```

### Threshold

`[threshold=10] {number}`

Pixel value before swiping will activate (helps on iOS to optimize scrolling).

> Warning! Creates a minor jump depending on the value.  
> 10px seems to be default for most developers and similar plugins.  

```js
import {multiBelt} from 'paraglider/presets'

multiBelt(document.querySelector('.belt'), {threshold: 5})
```

### Snap back

`[snapBackAt=0.25] {number}`

Progress needed to snap to the next slide. [0, 1]

```js
import {multiBelt} from 'paraglider/presets'

multiBelt(document.querySelector('.belt'), {snapBackAt: 0.5})
```

#### Speed and duration

`[speed=250] {number}`  

Speed defines the time the animation takes when `nextSlide` or `prevSlide` are called.  

`[spring=100] {number}`

Spring defines the time the animation takes when touch or mouse interaction is stopped (touchend, mouseup)

```js
import {belt} from 'paraglider/presets'

belt(document.querySelector('.belt'), {speed: 1000, spring: 300})
```

### Add classes

`addClasses {object}`  
* `previous {boolean}`
* `current {boolean}`
* `next {boolean}`

Allows defining if class names should be added to the slides.

`addMultiClasses {object}`  
* `previous {boolean}`
* `current {boolean}`
* `next {boolean}`

Allows defining if class names should be added to the slides if multiple are visible.

`current__0, current__1, ...`

```js
import {belt} from 'paraglider/presets'

belt(document.querySelector('.belt'), {addClasses: {current: true}})
```


## Callbacks

* `{(null|onInit)} [onInit=null]`
  * Callback when the slider has been created.
* `{(null|onDestroy)} [onDestroy=null]`
  * Callback when the slider has been destroyed.
* `{(null|onSlide)} [onSlide=null]`
  * Callback while the slider is moving.
* `{(null|onEnd)} [onEnd=null]`
  * Callback while the slider stopped moving.

### Initialized

`[onInit=null] {null|function}`  

Defaults to null and does nothing. Can be a function with parameters

```js
import Glider from 'paraglider'

const glider = new Glider({
  onInit({previous, current, next, rest}, slides, options) {
    slides[current].style.transform = 'translate3d(0,0,0)'
  }
})
```

### Destroyed

`[onDestroy=null] {null|function}`  

Defaults to null and does nothing. Can be a function

```js
import Glider from 'paraglider'

const glider = new Glider({
  onDestroy() {
    // ...
  }
})
```

### Sliding

`[onSlide=null] {null|function}`  

Defaults to null and does nothing. Can be a function with parameters

```js
import Glider from 'paraglider'

const glider = new Glider({
  onSlide(progress, {previous, current, next, rest}, slides, options) {
    slides[current].style.transform = 'translate3d(${100 * progress}%,0,0)'
    // ...
  }
})
```

### Finished

`[onEnd=null] {null|function}`  

Defaults to null and does nothing. Can be a function with parameters

```js
import Glider from 'paraglider'

const glider = new Glider({
  onEnd({previous, current, next, rest}, slides, options) {
    slides[current].style.transform = 'translate3d(0,0,0)'
  }
})
```

© 2017 by [Gregor Adams](greg@pixelass.com)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.0.0"></a>
# [3.0.0](https://github.com/pixelass/paraglider/compare/v2.4.0...v3.0.0) (2017-07-02)


### build

* **polyfills:** don't polyfill ([7358836](https://github.com/pixelass/paraglider/commit/7358836))


### Code Refactoring

* **dist:** dist is generated from main ([e831461](https://github.com/pixelass/paraglider/commit/e831461))
* **exports:** plugin only exports API ([48a0cef](https://github.com/pixelass/paraglider/commit/48a0cef))


### BREAKING CHANGES

* **dist:** Dist only exposes API. No presets or wrappers
* **exports:** No access to presets in main
* **polyfills:** no polyfills in dist versions



<a name="2.4.0"></a>
# [2.4.0](https://github.com/pixelass/paraglider/compare/v2.3.0...v2.4.0) (2017-07-02)


### Features

* **dist:** Added data-wrapper ([467f987](https://github.com/pixelass/paraglider/commit/467f987))
* **helpers:** preventdefault and parseObject ([44e4bd4](https://github.com/pixelass/paraglider/commit/44e4bd4))
* **methods:** onDestory exposes options ([cef0995](https://github.com/pixelass/paraglider/commit/cef0995))
* **methods:** return values on methods ([eb343e9](https://github.com/pixelass/paraglider/commit/eb343e9))
* **presets:** Added new wrapper ([06a0600](https://github.com/pixelass/paraglider/commit/06a0600))



<a name="2.3.0"></a>
# [2.3.0](https://github.com/pixelass/paraglider/compare/v2.2.0...v2.3.0) (2017-07-01)


### Features

* **options:** Added various options ([c56173d](https://github.com/pixelass/paraglider/commit/c56173d))



<a name="2.2.0"></a>
# [2.2.0](https://github.com/pixelass/paraglider/compare/v2.1.0...v2.2.0) (2017-06-28)


### Features

* **presets:** multi as wrapper ([53f2638](https://github.com/pixelass/paraglider/commit/53f2638))



<a name="2.1.0"></a>
# [2.1.0](https://github.com/pixelass/paraglider/compare/v2.0.4...v2.1.0) (2017-06-27)



<a name="2.0.4"></a>
## [2.0.4](https://github.com/pixelass/paraglider/compare/v2.0.3...v2.0.4) (2017-06-19)


### Bug Fixes

* **IE11:** support IE11 ([294a81c](https://github.com/pixelass/paraglider/commit/294a81c))



<a name="2.0.3"></a>
## [2.0.3](https://github.com/pixelass/paraglider/compare/v2.0.2...v2.0.3) (2017-06-16)


### Bug Fixes

* **wrapper:** pager context ([8cc51fe](https://github.com/pixelass/paraglider/commit/8cc51fe)), closes [#2](https://github.com/pixelass/paraglider/issues/2)



<a name="2.0.2"></a>
## [2.0.2](https://github.com/pixelass/paraglider/compare/v2.0.1...v2.0.2) (2017-06-13)


### Bug Fixes

* **defaults:** use correct default options ([081a0b5](https://github.com/pixelass/paraglider/commit/081a0b5))



<a name="2.0.1"></a>
## [2.0.1](https://github.com/pixelass/paraglider/compare/v2.0.0...v2.0.1) (2017-06-05)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/pixelass/paraglider/compare/v1.1.0...v2.0.0) (2017-06-04)


### Code Refactoring

* **API:** new API ([f7883ac](https://github.com/pixelass/paraglider/commit/f7883ac))


### BREAKING CHANGES

* **API:** API changed



<a name="1.1.0"></a>
# [1.1.0](https://github.com/pixelass/paraglider/compare/v1.0.0...v1.1.0) (2017-06-02)


### Bug Fixes

* **events:** prevent default events ([814eeca](https://github.com/pixelass/paraglider/commit/814eeca))


### Features

* **dist:** label dist as a feature ([5e5522a](https://github.com/pixelass/paraglider/commit/5e5522a))



<a name="1.0.0"></a>
# 1.0.0 (2017-06-02)


### Bug Fixes

* **presets:** direction on belt ([8aaecb5](https://github.com/pixelass/paraglider/commit/8aaecb5))


### Features

* **alpha:** alpha version of plugin ([00606be](https://github.com/pixelass/paraglider/commit/00606be))
* **presets:** added slider presets ([38461b1](https://github.com/pixelass/paraglider/commit/38461b1))


src/config.js
/**
 * Config data for Paraglider.
 *
 * @file config.js
 * @module config
 * @author Gregor Adams <greg@pixelass.com>
 */

/**
 * Default classList for the plugin.
 * This object can be replaced but not merged
 * @private
 * @type {object}
 * @property {string} pluginLoaded Applied when the plugin has been loaded
 * @property {string} init Applied when the plugin has been initialized. Removed on first interaction.
 * @property {string} slides This element will be used to track touches. This is the wrapper around the slides.
 * @property {string} slide Selector for each single slide.
 * @property {string} current Applied to the currently visible slide
 * @property {string} previous Applied to the previous slide in the queue
 * @property {string} next Applied to the next slide in the queue
 */
const classNames = {
  pluginLoaded: 'pluginLoaded',
  slide: 'slide',
  slides: 'slides',
  init: 'init',
  current: 'current',
  previous: 'previous',
  next: 'next'
}

/**
 * Defaults for the presets.
 * @type {object}
 * @property {object} [classNames]
 *   Mapping of class names to be used by the plugin.
 * @property {string} [classNames.pluginLoaded='pluginLoaded']
 *   Applied when the plugin has been loaded
 * @property {string} [classNames.init='init']
 *   Applied when the plugin has been initialized. Removed on first interaction.
 * @property {string} [classNames.slides='slides']
 *   This element will be used to track touches. This is the wrapper around the slides.
 * @property {string} [classNames.slide='slide']
 *   Selector for each single slide.
 * @property {string} [classNames.current='current']
 *   Applied to the currently visible slide
 * @property {string} [classNames.previous='previous']
 *   Applied to the previous slide in the queue
 * @property {string} [classNames.next='next']
 *   Applied to the next slide in the queue
 * @property {object} [addClasses]
 *   Map of class names to be added to slides.
 * @property {object} [addClasses.previous=true]
 *   Sets previous class name to slides.
 * @property {object} [addClasses.current=true]
 *   Sets current class name to slides.
 * @property {object} [addClasses.next=true]
 *   Sets next class name to slides.
 * @property {object} [addMultiClasses]
 *   Map of class names to be added to slides using a counter.
 * @property {object} [addMultiClasses.previous=undefined]
 *   Sets previous class name to slides using a counter.
 * @property {object} [addMultiClasses.current]
 *   Sets current class name to slides using a counter.
 * @property {object} [addMultiClasses.next=undefined]
 *   Sets next class name to slides using a counter.
 * @property {boolean} [enableTouch=true]
 *   Enable touch interaction.
 * @property {boolean} [enableSwipe=true]
 *   Enable mouse/drag/swipe interaction.
 * @property {boolean} [loop=false]
 *   Enable looping slides.
 * @property {?onInit} [onInit=null]
 *   Callback when the slider has been created.
 * @property {?onDestroy} [onDestroy=null]
 *   Callback when the slider has been destroyed.
 * @property {?onSlide} [onSlide=null]
 *   Callback while the slider is moving.
 * @property {?onEnd} [onEnd=null]
 *   Callback while the slider stopped moving.
 * @property {number} [speed=250]
 *   Animation duration when using paging.
 * @property {number} [spring=100]
 *   Animation duration when snapping.
 * @property {number} [snapBackAt=0.25]
 *   Amount of distance needed to snap. [0, 1]
 * @property {number} [threshold=10]
 *   Threshold of pixels until the sliding mechanisms is triggered.
 * @property {number} [initialSlide=0]
 *   Initially visible slide
 * @property {number} [visibleSlides=1]
 *   Amount of visible slides
 * @property {number} [slideBy=1]
 *   Amount of slides to slide on interaction
 */
const PLUGIN_DEFAULTS = {
  classNames,
  addClasses: {previous: true, current: true, next: true},
  addMultiClasses: {current: true},
  enableTouch: true,
  enableSwipe: true,
  loop: true,
  onInit: null,
  onDestroy: null,
  onSlide: null,
  onEnd: null,
  speed: 250,
  spring: 100,
  snapBackAt: 0.25,
  threshold: 10,
  initialSlide: 0,
  visibleSlides: 1,
  slideBy: 1
}

/**
 * Defaults for the presets.
 * @type {object}
 * @property {object} [classNames]
 *   Mapping of class names to be used by the plugin.
 * @property {string} [classNames.pluginLoaded='pluginLoaded']
 *   Applied when the plugin has been loaded
 * @property {string} [classNames.init='init']
 *   Applied when the plugin has been initialized. Removed on first interaction.
 * @property {string} [classNames.slides='slides']
 *   This element will be used to track touches. This is the wrapper around the slides.
 * @property {string} [classNames.slide='slide']
 *   Selector for each single slide.
 * @property {string} [classNames.current='current']
 *   Applied to the currently visible slide
 * @property {string} [classNames.previous='previous']
 *   Applied to the previous slide in the queue
 * @property {string} [classNames.next='next']
 *   Applied to the next slide in the queue
 * @property {string} [classNames.dot='dot']
 *   Selector for pager dots.
 * @property {string} [classNames.active='active']
 *   Active class for pager dots.
 * @property {string} [classNames.nextButton='nextButton']
 *   Selector for the navigation to the next slide.
 * @property {string} [classNames.prevButton='prevButton']
 *   Selector for the navigation to the previous slide.
 * @property {object} [addClasses]
 *   Map of class names to be added to slides.
 * @property {object} [addClasses.previous=true]
 *   Sets previous class name to slides.
 * @property {object} [addClasses.current=true]
 *   Sets current class name to slides.
 * @property {object} [addClasses.next=true]
 *   Sets next class name to slides.
 * @property {object} [addMultiClasses]
 *   Map of class names to be added to slides using a counter.
 * @property {object} [addMultiClasses.previous=undefined]
 *   Sets previous class name to slides using a counter.
 * @property {object} [addMultiClasses.current]
 *   Sets current class name to slides using a counter.
 * @property {object} [addMultiClasses.next=undefined]
 *   Sets next class name to slides using a counter.
 * @property {boolean} [enableTouch=true]
 *   Enable touch interaction.
 * @property {boolean} [enableSwipe=true]
 *   Enable mouse/drag/swipe interaction.
 * @property {boolean} [loop=false]
 *   Enable looping slides.
 * @property {?onInit} [onInit=null]
 *   Callback when the slider has been created.
 * @property {?onDestroy} [onDestroy=null]
 *   Callback when the slider has been destroyed.
 * @property {?onSlide} [onSlide=null]
 *   Callback while the slider is moving.
 * @property {?onEnd} [onEnd=null]
 *   Callback while the slider stopped moving.
 * @property {number} [speed=250]
 *   Animation duration when using paging.
 * @property {number} [spring=100]
 *   Animation duration when snapping.
 * @property {number} [snapBackAt=0.25]
 *   Amount of distance needed to snap. [0, 1]
 * @property {number} [threshold=10]
 *   Threshold of pixels until the sliding mechanisms is triggered.
 * @property {number} [initialSlide=0]
 *   Initially visible slide
 * @property {number} [visibleSlides=1]
 *   Amount of visible slides
 * @property {number} [slideBy=1]
 *   Amount of slides to slide on interaction
 */
const PRESET_DEFAULTS = {
  ...PLUGIN_DEFAULTS,
  classNames: {
    ...classNames,
    dot: 'dot',
    active: 'active',
    nextButton: 'nextButton',
    prevButton: 'prevButton'
  }
}

export {
  classNames,
  PLUGIN_DEFAULTS,
  PRESET_DEFAULTS
}
